### PR TITLE
add and integrate a SL2 in natural rep. recognition algorithm by Frank Lübeck 

### PR DESCRIPTION
**Pull request by Till-Eisen**
Added Frank Lübeck's RecogNaturalSL2 algorithm as well as ConRecogNaturalSL2, which calls it and converts the output to the same format as RECOG.RecogniseSL2NaturalEvenChar / RECOG.RecogniseSL2NaturalOddCharUsingBSGS. ConRecogNaturalSL2 retries up to 100 times until the correct generators are found (usually takes very few attempts, i.e. 100 is a very high bound).
Also added test_ConRecogNaturalSL2 for testing: it accepts either a list of prime powers or an empty list — in the latter case a preset list of prime powers is tested automatically.

Note RecogNaturalSL2 does not work for q=2,3,5. In this case the old functions are called.

### CODE REVIEW

File: gap/projective/constructive_recognition/SL/BaseCase.gi [hunk 1]
```diff
@@ -475,3 +475,355 @@ RECOG.FindStdGensUsingBSGS := function(g,stdgens,projective,large)
...
+  return [List([xm, trlowm, trupm], SLPOfElm), mat];
+end;
+
+## The following code is provided by Till Eisenbrand.
+## It integrates the algorithm above to produce the same output format
+## as RECOG.RecogniseSL2NaturalEvenChar and
+## RECOG.RecogniseSL2NaturalOddCharUsingBSGS.
+## G must be SL(2,q) generated by 2x2 matrices over GF(q).
+## Note: does not work for q = 2, 3, 5.
+ConRecogNaturalSL2 := function(G, f)
```
Comment (fingolfin) [inline gap/projective/constructive_recognition/SL/BaseCase.gi]:
```suggestion
RECOG.ConRecogNaturalSL2 := function(G, f)
```

File: gap/projective/constructive_recognition/SL/BaseCase.gi [hunk 2]
```diff
@@ -475,3 +475,355 @@ RECOG.FindStdGensUsingBSGS := function(g,stdgens,projective,large)
...
+      res_old := RECOG.RecogniseSL2NaturalEvenChar(G,f,false);
+    else
+      res_old := RECOG.RecogniseSL2NaturalOddCharUsingBSGS(G,f);
+    fi;
+    res := ConRecogNaturalSL2(G,f);
+    ## compare all generators after change of basis
+    for i in [1..Length(res.all)] do
+      if res.all[i]^res.basi <> res_old.all[i]^res_old.basi then
+        Print("Test failed for q = ", q, ", index i = ", i, " in the list \"all\" failed\n");
+            valid := false;
```
Comment (SoongNoonien) [inline gap/projective/constructive_recognition/SL/BaseCase.gi]:
```suggestion
        valid := false;
```
I think the indentation is still wrong here.

File: gap/projective/constructive_recognition/utils/achieve.gi [hunk 3]
```diff
@@ -870,10 +870,10 @@ function(ri, g)
...
       Info(InfoRecog,2,"ClassicalNatural: this is PSL_2!");
       if IsEvenInt(q) then
-          std := RECOG.RecogniseSL2NaturalEvenChar(gm,f,false);
+          std := RECOG.ConRecogNaturalSL2(gm,f);
           ri!.comment := "PSL2Even";
       else
-          std := RECOG.RecogniseSL2NaturalOddCharUsingBSGS(gm,f);
+          std := RECOG.ConRecogNaturalSL2(gm,f);
           ri!.comment := "PSL2Odd";
       fi;
```
Comment (fingolfin) [inline gap/projective/constructive_recognition/utils/achieve.gi]: This could be merged together now, with `ri!.comment := "PSL2";`

File: gap/projective/constructive_recognition/SL/BaseCase.gi [hunk 4]
```diff
@@ -475,3 +475,351 @@ RECOG.FindStdGensUsingBSGS := function(g,stdgens,projective,large)
...
+      m := u1^0;
+      for i in [0..j-1] do
+        c := IntFFE(coeffs[i+1]);
+        m := m * (diag^i * u1 * diag^(-i))^c;
+      od;
+      Add(lmat, m);
+    od;
+    umat := [];
+    for l in [0..j-1] do
+      coeffs := Coefficients(Basis(GF(q), basis), Z(q)^l);
```
Comment (fingolfin) [inline gap/projective/constructive_recognition/SL/BaseCase.gi]:
```suggestion
      coeffs := Coefficients(basis, Z(q)^l);
```
